exporter/prometheusremotewriteexporter/exporter.go

```diff
@@ -148,71 +150,105 @@ func validateAndSanitizeExternalLabels(cfg *Config) (map[string]string, error) {
+		go func(partition map[string]*prompb.TimeSeries, workerIndex int) {
 			defer wg.Done()
-			for {
-				select {
-				case <-ctx.Done(): // Check firstly to ensure that the context wasn't cancelled.
-					return
-
-				case request, ok := <-input:
-					if !ok {
-						return
-					}
-					if errExecute := prwe.execute(ctx, request); errExecute != nil {
-						mu.Lock()
-						errs = multierr.Append(errs, consumererror.NewPermanent(errExecute))
-						mu.Unlock()
-					}
+
+			requests, err := batchTimeSeries(partition, maxBatchByteSize)
+			if err != nil {
+				errorsArray[workerIndex] = consumererror.NewPermanent(err)
+				prwe.settings.Logger.Error("Failed to batch time series:", zap.Error(err))
+				return
+			}
+
+			if !prwe.walEnabled() {
+				// WAL is not enabled, perform a direct export
+				if errExecute := prwe.export(ctx, requests); errExecute != nil {
+					errorsArray[workerIndex] = errExecute
+					prwe.settings.Logger.Error("Failed to export data:", zap.Error(errExecute))
+				}
+			} else {
+				// WAL is enabled, persist requests to the WAL for later export
+				if err := prwe.wal.persistToWAL(requests); err != nil {
+					errorsArray[workerIndex] = consumererror.NewPermanent(err)
+					prwe.settings.Logger.Error("Failed to persist to WAL:", zap.Error(err))
 				}
 			}
-		}()
+		}(partitions[i], i)
 	}
+
+	// Wait for all workers to finish processing the requests
 	wg.Wait()
 
-	return errs
+	// Check if any errors were written to the errorsArray, and return them if present.
+	var finalErr error
+	for _, err := range errorsArray {
+		if err != nil {
+			finalErr = multierr.Append(finalErr, err)
+		}
+	}
+
+	return finalErr
+}
+
+// partitionTimeSeries partitions the time series map into N arrays using hashing/fnv.
+func partitionTimeSeries(tsMap map[string]*prompb.TimeSeries, concurrencyLimit int) []map[string]*prompb.TimeSeries {
+	partitions := make([]map[string]*prompb.TimeSeries, concurrencyLimit)
+	for i := 0; i < concurrencyLimit; i++ {
+		partitions[i] = make(map[string]*prompb.TimeSeries)
+	}
+
+	// FNV-1a is a non-cryptographic hash function.
+	hasher := fnv.New32a()
+	for key, ts := range tsMap {
+		hasher.Reset()
+		hasher.Write([]byte(key))
+		hash := hasher.Sum32()
+		// Modulo the key with the number of partitions to get the partition index.
+		partitionIndex := int(hash % uint32(concurrencyLimit))
+		partitions[partitionIndex][key] = ts
+	}
+	return partitions
+}
+
+func (prwe *prwExporter) export(ctx context.Context, requests []*prompb.WriteRequest) error {
+	// Process each WriteRequest sequentially
+	for _, request := range requests {
+		// Check if the context is already cancelled
+		select {
+		case <-ctx.Done():
+			// Even if the context is already cancelled, return the accumulated errors, if any.
+			return ctx.Err()
+		default:
+			// Process the WriteRequest
+			if errExecute := prwe.execute(ctx, request); errExecute != nil {
+				return consumererror.NewPermanent(errExecute)
```

Comment:
you need to collect all errors and only return in the end. similarly to what you are doing [here.](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24460/files#diff-2631d8a8855fdc08cf077c0ab99544147ca73eec572bc871d8fe717e826eda38R208)